rclcpp/test/rclcpp/test_node.cpp

```diff
@@ -334,11 +334,17 @@ TEST_F(TestNode, declare_parameter_with_no_initial_values) {
   {
     rcl_interfaces::msg::ParameterDescriptor descriptor;
     descriptor.dynamic_typing = true;
-    // no default, no initial
+    // no default, no initial, dynamic typing
     rclcpp::ParameterValue value = node->declare_parameter(
       "parameter"_unq, rclcpp::ParameterValue{}, descriptor);
     EXPECT_EQ(value.get_type(), rclcpp::PARAMETER_NOT_SET);
   }
+  {
+    // no default, no initial, static typing
+    rclcpp::ParameterValue value = node->declare_parameter(
+      "parameter"_unq, rclcpp::ParameterType::PARAMETER_INTEGER);
+    EXPECT_EQ(value.get_type(), rclcpp::PARAMETER_NOT_SET);
+  }
```

Comment:
it would be good to test what happens when you try to get the parameter.

IMO, if static typing is specified and we allow them to be uninitialized, the first get should throw an error if it's still uninitialized.
Does that make sense to you?

Reply:
We get an exception if you try to access a parameter that is uninitialized (static or dynamic). I've added tests and changed the code to throw a more accurate exceptions (01e5ae2).

I'm still not sure if we want to throw an exception if `dynamic_typing=true` or return an unset parameter. I think it's easy to implement the latter, but I went with throwing an exception for consistency.

Reply:
> I'm still not sure if we want to throw an exception if dynamic_typing=true or return an unset parameter. I think it's easy to implement the latter, but I went with throwing an exception for consistency.

mmm, in that case we should return the unset parameter IMO.
That was the behavior in previous releases for dynamic typing, and I would rather keep it so people can exactly get the previous behavior by using `dynamic_typing=false`.

Reply:
Another thing to test, you can declare an uninitialized statically typed parameter, then set it, then get it.
That should be valid without any exception thrown

Reply:
I suppose, though we technically the dynamic vs. static typing feature is new in Galactic, so I don't think there's many people using it yet. Though, I don't mind making the behavior such that it returns NOT_SET

Reply:
Test added in 344149c
Not longer throwing exception in the dynamic typing case: 1ea6ea4